x-pack/plugin/core/src/main/java/org/elasticsearch/index/engine/FrozenEngine.java

```diff
@@ -0,0 +1,520 @@
+import org.apache.lucene.index.FilterLeafReader;
+import org.apache.lucene.index.IndexCommit;
+import org.apache.lucene.index.IndexReader;
+import org.apache.lucene.index.IndexWriter;
+import org.apache.lucene.index.LeafMetaData;
+import org.apache.lucene.index.LeafReader;
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.index.NumericDocValues;
+import org.apache.lucene.index.PointValues;
+import org.apache.lucene.index.SegmentCommitInfo;
+import org.apache.lucene.index.SortedDocValues;
+import org.apache.lucene.index.SortedNumericDocValues;
+import org.apache.lucene.index.SortedSetDocValues;
+import org.apache.lucene.index.StoredFieldVisitor;
+import org.apache.lucene.index.Terms;
+import org.apache.lucene.search.IndexSearcher;
+import org.apache.lucene.search.ReferenceManager;
+import org.apache.lucene.store.AlreadyClosedException;
+import org.apache.lucene.util.Bits;
+import org.elasticsearch.common.SuppressForbidden;
+import org.elasticsearch.common.lucene.Lucene;
+import org.elasticsearch.core.internal.io.IOUtils;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.List;
+import java.util.function.Function;
+
+/**
+ * This is a stand-alone read-only engine that maintains a lazy loaded index reader that is opened on calls to
+ * {@link Engine#acquireSearcher(String)}. The index reader opened is maintained until there are no reference to it anymore and then
+ * releases itself from the engine. The readers returned from this engine are lazy which allows release after and reset before a search
+ * phase starts. This allows releasing references as soon as possible on the search layer.
+ *
+ * Internally this class uses a set of wrapper abstractions to allow a reader that is used inside the {@link Engine.Searcher} returned from
+ * {@link #acquireSearcher(String, SearcherScope)} to release and reset it's internal resources. This is necessary to for instance release
+ * all SegmentReaders after a search phase finishes and reopen them before the next search phase starts. This together with a throttled
+ * threadpool (search_throttled) guarantees that at most N frozen shards have a low level index reader open at the same time.
+ *
+ * In particular we have LazyDirectoryReader that wraps its LeafReaders (the actual segment readers) inside LazyLeafReaders. Each of the
+ * LazyLeafReader delegates to segment LeafReader that can be reset (it's reference decremented and nulled out) on a search phase is
+ * finished. Before the next search phase starts we can reopen the corresponding reader and reset the reference to execute the search phase.
+ * This allows the SearchContext to hold on to the same LazyDirectoryReader across its lifecycle but under the hood resources (memory) is
+ * released while the SearchContext phases are not executing.
+ *
+ * The internal reopen of readers is treated like a refresh and refresh listeners are called up-on reopen. This allows to consume refresh
```

Comment:
where is this happening?

Reply:
see https://github.com/elastic/elasticsearch/pull/34357/files#diff-fa7c1eacb2e5cd3b73e133d9d35d2342R135

Reply:
yep. sorry I missed it.